### PR TITLE
ToQUBO Compiler Upgrade

### LINKED ISSUE #69: `NTR_KZFD` quadratization dispatch not being called
This function should verify if the coefficient for each term is negative, so it can call the `NTR_KZFD` dispatch.

https://github.com/psrenergy/ToQUBO.jl/blob/0a7b10bd938a93e98eb97b5deaa3e52e929fd7be/src/lib/pbo/quadratization.jl#L132-L148

We can just follow the same logic from this dispatch:

https://github.com/psrenergy/ToQUBO.jl/blob/0a7b10bd938a93e98eb97b5deaa3e52e929fd7be/src/lib/pbo/quadratization.jl#L116-L130

### LINKED ISSUE #75: Place `toqubo_*` methods inside new `Compiler` module
- [ ] Create `ToQUBO.Compiler` module
- [ ] Move `toqubo_*` function definitions to it

### LINKED ISSUE #78: Enhance and Validate the Attribute System
This involves adding tests for attribute definition and retrieval, specially with respect to the interface with JuMP.

**NOTE:** It is important to leverage jump-dev/MathOptInterface.jl#2331 when dealing with variable attributes.

### LINKED ISSUE #79: Add hint attributes for penalty factors
This issue is a follow-up on the discussion at jump-dev/MathOptInterface.jl#2345.

- [x] `TQA.ConstraintEncodingPenaltyHint`
- [x] `TQA.VariableEncodingPenaltyHint`
- [x] Add tests for both

### PR DESCRIPTION
**Pull request by pedromxavier**
This PR aims to make the reformulation modular, helping the development and testing of its current and subsequent features.

The necessary tasks to merge it are:
- [x] Fix the quadratization issues (closes #69)
- [x] Establish the `Compiler` module (closes #75)
- [x] Consolidate the Attribute System (closes #78, closes #79)

### CODE REVIEW

File: src/virtual/model.jl [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+            Dict{VI,Variable{T}}(),
+            Dict{VI,Variable{T}}(),
+
+            #  PBO/PBF IR 
+            PBO.PBF{VI,T}(),          # Objective Function
+            Dict{CI,PBO.PBF{VI,T}}(), # Constraint Functions
+            Dict{VI,PBO.PBF{VI,T}}(), # Variable Functions
+            Dict{CI,T}(),             # Constraint Penalties
+            Dict{VI,T}(),             # Variable Penalties
+            PBO.PBF{VI,T}(),          # Final Hamiltonian
```
Comment (bernalde) [inline src/virtual/model.jl:70]: Let's decide what to name the final Q, Hamiltonian is almost (but not quite) correct

File: test/runtests.jl [hunk 2]
```diff
@@ -4,27 +4,39 @@ using JuMP
...
+end
 
-include("assets/assets.jl")
 include("unit/unit.jl")
 include("integration/integration.jl")
-include("examples/examples.jl")
 
 function main()
-    @testset "::  ::  :: ToQUBO.jl ::  ::  ::" verbose = true begin
+    @testset "♡ ToQUBO.jl $(ToQUBO.__VERSION__) Test Suite ♡" verbose = true begin
```
Comment (bernalde) [inline test/runtests.jl:34]: Lol

File: test/unit/encoding/encoding.jl [hunk 3]
```diff
@@ -0,0 +1,9 @@
+include("variables.jl")
+
+function test_encoding_methods()
+    @testset "□ Encoding" verbose = true begin
+        test_variable_encoding_methods()
+    end
+
+    return nothing
+end
```
Comment (bernalde) [inline test/unit/encoding/encoding.jl:9]: Missing end line

**Review by bernalde — APPROVED** [review]
Massive PR with many improvements. Good work. Many of the tests, if I understand correctly, were just moved to other packages.
My comments are really minor. Well done Pedro!

File: src/compiler/build.jl [hunk 4]
```diff
@@ -42,43 +43,47 @@ function toqubo_hamiltonian!(model::VirtualModel{T}, ::AbstractArchitecture) whe
...
+            # NOTE: Here it is necessary to invert the sign of the
+            # Hamiltonian since PBO adopts the minimization sense
+            # convention.
+            # TODO: Add an in-place version of 'quadratize!' that 
+            # provides support for maximization problems.
+            let H = PBO.quadratize!(-model.H, quad) do (n::Union{Integer,Nothing} = nothing)
+                    return aux(model, n, arch)
+                end
+
+                # NOTE: This setup leads to avoidable allocations.
```
[the remark this answers is not in the document]
Comment (pedromxavier) [reply src/compiler/build.jl:73]: Basically, the quadratization process would need to be aware of the optimization sense in some way to perform the correct in-place operations.

[the remark this answers is not in the document]
Comment (pedromxavier) [reply src/compiler/build.jl]: As a quick fix, I've negated H in-place, before and after

File: src/virtual/model.jl (hunk 1 above)
Comment (pedromxavier) [reply src/virtual/model.jl]: Comments were updated

**Review by bernalde — APPROVED** [review]

**Pull request merged**